File: packages/host/src/board/SessionBoard.ts
/**
 * SessionBoard —— host 的「状态总表」。
 *
 * 所有 backend 把观察到的会话快照推给它，它负责：
 *   1. 合并所有 backend 的快照
 *   2. 调用 slotAllocator 决定哪个会话占哪个槽
 *   3. 调用 theme.ts 给每个槽位算出 RGB/亮度/fx
 *   4. 输出最终的 LedFrame（推给设备）+ BoardState（推给 simulator UI）
 *
 * 对外是不可变的数据流：通过 onLedFrame / onBoardState 订阅输出。
 */

import {
  type BackendId,
  type BoardState,
  type LedFrame,
  type LedSlot,
  type PolicyMode,
  type SessionSnapshot,
  type SlotBinding,
  DONE_TTL_MS,
  RISK_BOOST,
  SLOT_COUNT,
} from '@agent-deck/protocol'

import { bus } from '../bus.js'
import type { AgentBackend } from '../backends/types.js'
import {
  allocateSlots,
  type AllocatedSlot,
  type ScoredSession,
} from './slotAllocator.js'
import { CODEX_THEME, paint, type ThemePalette } from './theme.js'

// ─────────────────────────────────────────────────────────────────────────────
// 构造参数
// ─────────────────────────────────────────────────────────────────────────────

export interface SessionBoardOptions {
  /** 槽位数，默认 SLOT_COUNT */
  slotCount?: number
  /** 灯效调色板，默认 CODEX_THEME */
  palette?: ThemePalette
  /** 当前时间函数，默认 Date.now（测试可注入） */
  now?: () => number
}

// ─────────────────────────────────────────────────────────────────────────────
// 内部辅助
// ─────────────────────────────────────────────────────────────────────────────

interface BackendEntry {
  backend: AgentBackend
  unsubscribe: () => void
}

/**
 * 计算单个 session 的 urgency ∈ [0, 1]。
 *
 *   - 非 waiting 状态：urgency = 0
 *   - waiting：timeUrgency = clamp01(ageSec / 120)，再与 RISK_BOOST[risk] 取 max
 *
 * 与 theme.ts 内部使用的语义保持一致；这里独立实现避免循环依赖。
 */
function computeUrgency(snapshot: SessionSnapshot, now: number): number {
  if (snapshot.status !== 'waiting') return 0
  const waitingSince = snapshot.waitingSince
  if (waitingSince === undefined) return 0
  const ageSec = (now - waitingSince) / 1000
  const timeUrgency = Math.min(Math.max(ageSec / 120, 0), 1)
  const riskBoost = snapshot.risk ? (RISK_BOOST[snapshot.risk] ?? 0) : 0
  return Math.max(timeUrgency, riskBoost)
}

/** session 在 sessions Map 中的复合 key */
function sessionKey(backend: BackendId, sessionId: string): string {
  return `${backend}:${sessionId}`
}

// ─────────────────────────────────────────────────────────────────────────────
// SessionBoard
// ─────────────────────────────────────────────────────────────────────────────

export class SessionBoard {
  private readonly slotCount: number
  private readonly palette: ThemePalette
  private readonly now: () => number

  /** 所有 backend 推上来的 session 快照，key = `${backend}:${sessionId}` */
  private readonly sessions = new Map<string, SessionSnapshot>()

  /** 已注册的 backend，key = backend.id */
  private readonly backends = new Map<BackendId, BackendEntry>()

  /** 当前焦点槽 */
  private focus = 0

  /** 钉死的槽位 → sessionId（手动绑定，透传给 slotAllocator） */
  private readonly pins = new Map<number, string>()

  /** 策略模式（V1 不动态切换，默认 act） */
  private readonly mode: PolicyMode = 'act'

  /** led 帧订阅者 */
  private readonly ledHandlers = new Set<(frame: LedFrame) => void>()

  /** board state 订阅者 */
  private readonly stateHandlers = new Set<(state: BoardState) => void>()

  /** 防止 dispose 后再广播 */
  private disposed = false

  /** 最后一次输出的 led 帧（供新连接 client 立即取） */
  private lastLedFrame: LedFrame | null = null

  /** 最后一次输出的 board state（供新连接 client 立即取） */
  private lastBoardState: BoardState | null = null

  constructor(opts?: SessionBoardOptions) {
    this.slotCount = opts?.slotCount ?? SLOT_COUNT
    this.palette = opts?.palette ?? CODEX_THEME
    this.now = opts?.now ?? (() => Date.now())
  }

  // ─── backend 注册 ────────────────────────────────────────────────────────

  /** 注册一个 backend，开始订阅它的快照。 */
  addBackend(backend: AgentBackend): void {
    const id = backend.id
    // 已注册过：先解除旧订阅，保持幂等
    const existing = this.backends.get(id)
    if (existing) {
      existing.unsubscribe()
    }

    const unsubscribe = backend.observe((snapshots) => {
      this.handleSnapshots(id, snapshots)
    })

    this.backends.set(id, { backend, unsubscribe })
  }

  /** 移除 backend，清掉它的所有 session。 */
  removeBackend(id: BackendId): void {
    const entry = this.backends.get(id)
    if (!entry) return
    try {
      entry.unsubscribe()
    } catch (err) {
      // backend unsubscribe 不应抛，但兜底防止影响主流程
      this.reportError(err)
    }
    this.backends.delete(id)

    // 删除该 backend 的所有 session
    for (const key of Array.from(this.sessions.keys())) {
      if (key.startsWith(`${id}:`)) {
        this.sessions.delete(key)
      }
    }

    // 注：指向已消失 session 的 pin 不在此处清理 —— slotAllocator 在 session
    // 不存在时会把 pinned 槽留空但仍占位，正是 pin 的预期语义（防被抢占）。
    this.recompute()
  }

  /** backend observe 回调：用新列表覆盖该 backend 的所有 session。 */
  private handleSnapshots(backend: BackendId, snapshots: SessionSnapshot[]): void {
    if (this.disposed) return

    // 1. 删除该 backend 下、不在新列表里的 session
    const incoming = new Set<string>()
    for (const s of snapshots) {
      incoming.add(sessionKey(backend, s.sessionId))
    }
    for (const key of Array.from(this.sessions.keys())) {
      if (key.startsWith(`${backend}:`) && !incoming.has(key)) {
        this.sessions.delete(key)
      }
    }

    // 2. 更新 / 插入新列表中的 session
    for (const s of snapshots) {
      this.sessions.set(sessionKey(backend, s.sessionId), s)
    }

    // 3. 触发重新计算
    this.recompute()
  }

  // ─── 用户操作 ──────────────────────────────────────────────────────────────

  /** 设置当前焦点槽。 */
  setFocus(i: number): void {
    if (i < 0 || i >= this.slotCount || !Number.isInteger(i)) return
    if (this.focus === i) return
    this.focus = i
    this.recompute()
  }

  /** 设置 / 取消钉死。sessionId 为 null 表示取消该槽位的 pin。 */
  pin(i: number, sessionId: string | null): void {
    if (i < 0 || i >= this.slotCount || !Number.isInteger(i)) return
    if (sessionId === null) {
      if (this.pins.delete(i)) {
        this.recompute()
      }
      return
    }
    const prev = this.pins.get(i)
    if (prev === sessionId) return
    this.pins.set(i, sessionId)
    this.recompute()
  }

  // ─── 重新计算 ────────────────────────────────────────────────────────────

  /** 强制重新计算并广播（外部触发或内部状态变更后调用）。 */
  recompute(): void {
    if (this.disposed) return

    const now = this.now()

    // 1. 扫一遍：过期的 done 直接从 sessions 删除
    this.purgeExpiredDone(now)

    // 2. 构造 ScoredSession[]
    const scored: ScoredSession[] = []
    for (const s of this.sessions.values()) {
      scored.push({ ...s, urgency: computeUrgency(s, now) })
    }

    // 3. 分配槽位
    const allocated = allocateSlots(scored, {
      slotCount: this.slotCount,
      focus: this.focus,
      pins: this.pins,
    })

    // 4. 组装 LedFrame + BoardState
    const ledFrame = this.buildLedFrame(allocated)
    const boardState = this.buildBoardState(allocated)

    // 5. 广播给订阅者
    this.broadcast(ledFrame, boardState)
  }

  /** 删除 status==='done' 且 now - updatedAt > DONE_TTL_MS 的 session。 */
  private purgeExpiredDone(now: number): void {
    for (const [key, s] of this.sessions) {
      if (s.status === 'done' && now - s.updatedAt > DONE_TTL_MS) {
        this.sessions.delete(key)
      }
    }
  }

  /** 由 AllocatedSlot[] 构造 LedFrame。 */
  private buildLedFrame(allocated: AllocatedSlot[]): LedFrame {
    const slots: LedSlot[] = allocated.map((slot) => {
      if (!slot.session) {
        return { i: slot.i, rgb: null, br: 0, fx: 'solid' as const }
      }
      const out = paint(
        {
          status: slot.session.status,
          risk: slot.session.risk,
          waitingSince: slot.session.waitingSince,
          now: this.now(),
        },
        this.palette,
      )
      return { i: slot.i, rgb: out.rgb, br: out.br, fx: out.fx }
    })
    return { type: 'leds', slots }
  }

  /** 由 AllocatedSlot[] 构造 BoardState。 */
  private buildBoardState(allocated: AllocatedSlot[]): BoardState {
    const slots: SlotBinding[] = allocated.map((slot) => ({
      i: slot.i,
      backend: slot.session?.backend,
      sessionId: slot.session?.sessionId,
      title: slot.session?.title,
      status: slot.session?.status ?? 'off',
      detail: slot.session?.detail,
      focused: slot.i === this.focus,
    }))
    return {
      type: 'board',
      slots,
      focus: this.focus,
      mode: this.mode,
    }
  }

  /** 广播 led 帧 + board state 给本地订阅者，并 emit 到 bus。 */
  private broadcast(ledFrame: LedFrame, boardState: BoardState): void {
    this.lastLedFrame = ledFrame
    this.lastBoardState = boardState
    for (const cb of this.ledHandlers) {
      try {
        cb(ledFrame)
      } catch (err) {
        this.reportError(err)
      }
    }
    for (const cb of this.stateHandlers) {
      try {
        cb(boardState)
      } catch (err) {
        this.reportError(err)
      }
    }
    try {
      bus.emit('board.changed', ledFrame)
    } catch (err) {
      this.reportError(err)
    }
  }

  // ─── 订阅 ──────────────────────────────────────────────────────────────────

  /** 订阅 led 帧（给 device bridge 用）。返回 unsubscribe。 */
  onLedFrame(cb: (frame: LedFrame) => void): () => void {
    this.ledHandlers.add(cb)
    return () => {
      this.ledHandlers.delete(cb)
    }
  }

  /** 订阅 board state（给 simulator gateway 用）。返回 unsubscribe。 */
  onBoardState(cb: (state: BoardState) => void): () => void {
    this.stateHandlers.add(cb)
    return () => {
      this.stateHandlers.delete(cb)
    }
  }

  // ─── Getter（外部读取当前状态，main.ts / ActionRouter 用） ──────────────────

  /** 当前焦点槽 */
  getFocus(): number {
    return this.focus
  }

  /** 最后一次推的 LedFrame（可能为 null） */
  getLedFrame(): LedFrame | null {
    return this.lastLedFrame
  }

  /** 最后一次推的 BoardState（可能为 null） */
  getBoardState(): BoardState | null {
    return this.lastBoardState
  }

  /** 当前策略模式 */
  getMode(): PolicyMode {
    return this.mode
  }

  // ─── 销毁 ──────────────────────────────────────────────────────────────────

  /** 销毁：清理所有订阅、清空 handlers。 */
  dispose(): void {
    if (this.disposed) return
    this.disposed = true

    for (const entry of this.backends.values()) {
      try {
        entry.unsubscribe()
      } catch (err) {
        this.reportError(err)
      }
    }
    this.backends.clear()
    this.sessions.clear()
    this.pins.clear()
    this.ledHandlers.clear()
    this.stateHandlers.clear()
  }

  // ─── 内部工具 ────────────────────────────────────────────────────────────────

  /** 统一错误出口（避免 console.log，仅记录真正的异常）。 */
  private reportError(err: unknown): void {
    // 仅在真正异常时输出；正常控制流不经过这里
    console.error('[SessionBoard] handler threw:', err)
  }
}

File: crates/board/src/session_board.rs
//! SessionBoard — merges backends, allocates slots, paints LEDs.
//! Ported from packages/host/src/board/SessionBoard.ts

use crate::slot_allocator::{allocate_slots, AllocatedSlot, ScoredSession, SlotAllocatorOptions};
use crate::theme::{paint, ThemeInput, ThemePalette, CODEX_THEME};
use agent_deck_protocol::{
    BackendId, BoardState, DeckStatus, LedFrame, LedSlot, PolicyMode, SessionSnapshot, SlotBinding,
    SLOT_COUNT,
};
use std::collections::HashMap;

pub struct SessionBoard {
    slot_count: usize,
    palette: ThemePalette,
    sessions: HashMap<String, SessionSnapshot>,
    focus: usize,
    pins: HashMap<usize, String>,
    /// When false, only pinned sessions occupy slots (manual bind mode).
    auto_fill: bool,
    mode: PolicyMode,
    last_led: Option<LedFrame>,
    last_board: Option<BoardState>,
}

impl Default for SessionBoard {
    fn default() -> Self {
        Self::new(SLOT_COUNT)
    }
}

impl SessionBoard {
    pub fn new(slot_count: usize) -> Self {
        Self {
            slot_count,
            palette: CODEX_THEME,
            sessions: HashMap::new(),
            focus: 0,
            pins: HashMap::new(),
            // Algorithm default is auto-fill; DesktopService may disable it for
            // manual-bind UX (empty keys until the user pins a session).
            auto_fill: true,
            mode: PolicyMode::Act,
            last_led: None,
            last_board: None,
        }
    }

    pub fn set_auto_fill(&mut self, enabled: bool, now: u64) {
        self.auto_fill = enabled;
        self.recompute(now);
    }

    pub fn auto_fill(&self) -> bool {
        self.auto_fill
    }

    /// All known sessions across backends (board cache; prefer catalog for bind UI).
    pub fn list_sessions(&self) -> Vec<SessionSnapshot> {
        let mut out: Vec<SessionSnapshot> = self.sessions.values().cloned().collect();
        out.sort_by(|a, b| {
            b.status
                .priority()
                .cmp(&a.status.priority())
                .then_with(|| b.updated_at.cmp(&a.updated_at))
        });
        out
    }

    /// Insert/replace a session snapshot without wiping the rest of the backend.
    /// Used when the user binds a historical session from the catalog.
    pub fn upsert_session(&mut self, snapshot: SessionSnapshot, now: u64) {
        let k = Self::key(snapshot.backend, &snapshot.session_id);
        self.sessions.insert(k, snapshot);
        self.recompute(now);
    }

    /// Insert/replace many snapshots at once, recomputing only once at the end.
    /// Used by the pinned-refresh path in `tick_at` to override stale cached
    /// rows with fresh state for bound sessions.
    pub fn upsert_sessions(&mut self, snapshots: Vec<SessionSnapshot>, now: u64) {
        if snapshots.is_empty() {
            return;
        }
        for snap in snapshots {
            let k = Self::key(snap.backend, &snap.session_id);
            self.sessions.insert(k, snap);
        }
        self.recompute(now);
    }

    pub fn find_session(&self, session_id: &str) -> Option<&SessionSnapshot> {
        self.sessions.values().find(|s| s.session_id == session_id)
    }

    fn key(backend: BackendId, session_id: &str) -> String {
        format!(
            "{}:{session_id}",
            match backend {
                BackendId::Zcode => "zcode",
                BackendId::Codex => "codex",
            }
        )
    }

    fn is_pinned_session(&self, session_id: &str) -> bool {
        self.pins.values().any(|id| id == session_id)
    }

    pub fn set_focus(&mut self, i: usize) {
        if i < self.slot_count {
            self.focus = i;
        }
    }

    pub fn focus(&self) -> usize {
        self.focus
    }

    /// Pin a session to slot `i`, or unpin it when `session_id` is None.
    /// Pinned slots reserve their position across recompute even if the session
    /// disappears, so a task stays bound to a fixed key (the "memory point").
    pub fn set_pin(&mut self, i: usize, session_id: Option<String>, now: u64) {
        if i >= self.slot_count {
            return;
        }
        match session_id {
            Some(id) => {
                self.pins.insert(i, id);
            }
            None => {
                self.pins.remove(&i);
            }
        }
        self.recompute(now);
    }

    /// Current pin map (slot index → session id), for persistence.
    pub fn pins(&self) -> &HashMap<usize, String> {
        &self.pins
    }

    /// Replace the whole pin map (e.g. when loading from disk on startup).
    pub fn set_pins(&mut self, pins: HashMap<usize, String>, now: u64) {
        self.pins = pins;
        self.recompute(now);
    }

    pub fn replace_backend_sessions(
        &mut self,
        backend: BackendId,
        snapshots: Vec<SessionSnapshot>,
        now: u64,
    ) {
        let prefix = match backend {
            BackendId::Zcode => "zcode:",
            BackendId::Codex => "codex:",
        };
        // Preserve manually pinned historical sessions even if they fall outside
        // the board poll window (LIMIT 20 / Done TTL / idle filter).
        let pinned_keep: HashMap<String, SessionSnapshot> = self
            .sessions
            .iter()
            .filter(|(k, s)| k.starts_with(prefix) && self.is_pinned_session(&s.session_id))
            .map(|(k, s)| (k.clone(), s.clone()))
            .collect();

        self.sessions.retain(|k, _| !k.starts_with(prefix));
        for s in snapshots {
            let k = Self::key(backend, &s.session_id);
            self.sessions.insert(k, s);
        }
        for (k, s) in pinned_keep {
            self.sessions.entry(k).or_insert(s);
        }
        self.recompute(now);
    }

    pub fn recompute(&mut self, now: u64) {
        // Done sessions are no longer purged by age — users want completed
        // sessions to stay visible. The slot allocator ranks Done lowest, so
        // active sessions always take precedence and Done only fills leftover
        // slots. (Pinned sessions are of course never dropped.)
        let _ = now;
        let pinned_ids: std::collections::HashSet<String> = self.pins.values().cloned().collect();
        self.sessions.retain(|_, s| {
            if pinned_ids.contains(&s.session_id) {
                return true;
            }
            // Only drop sessions that the backend itself stopped reporting
            // (handled by replace_backend_sessions); keep everything we have.
            true
        });

        let scored: Vec<ScoredSession> = self
            .sessions
            .values()
            .map(|s| ScoredSession {
                urgency: compute_urgency(s, now),
                snapshot: s.clone(),
            })
            .collect();

        let opts = SlotAllocatorOptions {
            slot_count: self.slot_count,
            focus: Some(self.focus),
            pins: self.pins.clone(),
        };
        let mut allocated = allocate_slots(&scored, &opts, now);
        // Manual mode: clear non-pinned slots so unbound keys stay Off.
        if !self.auto_fill {
            for slot in &mut allocated {
                if !slot.pinned {
                    slot.session = None;
                }
            }
        }
        let led = self.build_led_frame(&allocated, now);
        let board = self.build_board_state(&allocated);
        self.last_led = Some(led);
        self.last_board = Some(board);
    }

    fn build_led_frame(&self, allocated: &[AllocatedSlot], now: u64) -> LedFrame {
        let slots = allocated
            .iter()
            .map(|slot| {
                if let Some(ref session) = slot.session {
                    let out = paint(
                        &ThemeInput {
                            status: session.snapshot.status,
                            risk: session.snapshot.risk,
                            waiting_since: session.snapshot.waiting_since,
                            now,
                        },
                        &self.palette,
                    );
                    LedSlot {
                        i: slot.i,
                        rgb: out.rgb,
                        br: out.br,
                        fx: out.fx,
                    }
                } else {
                    LedSlot {
                        i: slot.i,
                        rgb: None,
                        br: 0,
                        fx: agent_deck_protocol::LedFx::Solid,
                    }
                }
            })
            .collect();
        LedFrame::new(slots)
    }

    fn build_board_state(&self, allocated: &[AllocatedSlot]) -> BoardState {
        let slots = allocated
            .iter()
            .map(|slot| {
                if let Some(ref session) = slot.session {
                    SlotBinding {
                        i: slot.i,
                        backend: Some(session.snapshot.backend),
                        session_id: Some(session.snapshot.session_id.clone()),
                        title: Some(session.snapshot.title.clone()),
                        status: session.snapshot.status,
                        detail: session.snapshot.detail.clone(),
                        focused: Some(slot.i == self.focus),
                        pinned: Some(slot.pinned),
                    }
                } else {
                    SlotBinding {
                        i: slot.i,
                        backend: None,
                        session_id: None,
                        title: None,
                        status: DeckStatus::Off,
                        detail: None,
                        focused: Some(slot.i == self.focus),
                        pinned: Some(slot.pinned),
                    }
                }
            })
            .collect();
        BoardState::new(slots, self.focus, self.mode)
    }

    pub fn led_frame(&self) -> Option<&LedFrame> {
        self.last_led.as_ref()
    }

    pub fn board_state(&self) -> Option<&BoardState> {
        self.last_board.as_ref()
    }
}

fn compute_urgency(snapshot: &SessionSnapshot, now: u64) -> f64 {
    if snapshot.status != DeckStatus::Waiting {
        return 0.0;
    }
    let Some(waiting_since) = snapshot.waiting_since else {
        return 0.0;
    };
    let age_sec = (now.saturating_sub(waiting_since)) as f64 / 1000.0;
    let time_urgency = (age_sec / 120.0).clamp(0.0, 1.0);
    let risk_boost = snapshot.risk.map(|r| r.boost()).unwrap_or(0.0);
    time_urgency.max(risk_boost)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn set_pin_reserves_slot_and_marks_board() {
        let mut board = SessionBoard::new(5);
        board.replace_backend_sessions(
            BackendId::Zcode,
            vec![
                SessionSnapshot {
                    backend: BackendId::Zcode,
                    session_id: "a".into(),
                    title: "A".into(),
                    status: DeckStatus::Working,
                    risk: None,
                    detail: None,
                    waiting_since: None,
                    updated_at: 100,
                    workspace_path: None,
                },
                SessionSnapshot {
                    backend: BackendId::Zcode,
                    session_id: "b".into(),
                    title: "B".into(),
                    status: DeckStatus::Done,
                    risk: None,
                    detail: None,
                    waiting_since: None,
                    updated_at: 200,
                    workspace_path: None,
                },
            ],
            1000,
        );
        // Pin the lower-priority "b" into slot 0 — it must stay there.
        board.set_pin(0, Some("b".into()), 1000);
        let state = board.board_state().unwrap();
        assert_eq!(state.slots[0].session_id.as_deref(), Some("b"));
        assert_eq!(state.slots[0].pinned, Some(true));
        assert_eq!(state.slots[1].session_id.as_deref(), Some("a"));
        assert_eq!(state.slots[1].pinned, Some(false));

        // Unpin restores free allocation.
        board.set_pin(0, None, 1000);
        let state = board.board_state().unwrap();
        assert_eq!(state.slots[0].pinned, Some(false));
        assert_eq!(state.slots[0].session_id.as_deref(), Some("a"));
    }

    #[test]
    fn working_session_paints_blue() {
        let mut board = SessionBoard::new(5);
        board.replace_backend_sessions(
            BackendId::Zcode,
            vec![SessionSnapshot {
                backend: BackendId::Zcode,
                session_id: "s1".into(),
                title: "t".into(),
                status: DeckStatus::Working,
                risk: None,
                detail: None,
                waiting_since: None,
                updated_at: 1000,
                workspace_path: None,
            }],
            1000,
        );
        let led = board.led_frame().unwrap();
        let occupied = led.slots.iter().find(|s| s.rgb.is_some()).unwrap();
        assert_eq!(occupied.fx, agent_deck_protocol::LedFx::Breathe);
        let rgb = occupied.rgb.unwrap();
        assert!(rgb[2] > 200);
    }
}
